# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [7]:
import pandas as pd
url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'
df = pd.read_csv(url)

In [8]:
df.head()

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [9]:
#Identify the dimensions of the dataset by determining the number of rows and columns it contains.
df.shape

(4008, 11)

In [10]:
#Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. 
df.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

In [11]:
df.sample(10)

Customer          ST GENDER             Education  \
2343      NaN         NaN    NaN                   NaN   
2556      NaN         NaN    NaN                   NaN   
3053      NaN         NaN    NaN                   NaN   
1101      NaN         NaN    NaN                   NaN   
305   QD28391  Washington      M               College   
1608      NaN         NaN    NaN                   NaN   
2506      NaN         NaN    NaN                   NaN   
263   GU66096     Arizona      M  High School or Below   
1099      NaN         NaN    NaN                   NaN   
2548      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
2343                     NaN      NaN                   NaN   
2556                     NaN      NaN                   NaN   
3053                     NaN      NaN                   NaN   
1101                     NaN      NaN                   NaN   
305               742584.53%  84302.0                  62.0   
1608                     NaN      NaN                   NaN   
2506                     NaN      NaN                   NaN   
263               545734.26%  94731.0                  67.0   
1099                     NaN      NaN                   NaN   
2548                     NaN      NaN                   NaN   

     Number of Open Complaints    Policy Type  Vehicle Class  \
2343                       NaN            NaN            NaN   
2556                       NaN            NaN            NaN   
3053                       NaN            NaN            NaN   
1101                       NaN            NaN            NaN   
305                     1/0/00  Personal Auto  Four-Door Car   
1608                       NaN            NaN            NaN   
2506                       NaN            NaN            NaN   
263                     1/0/00  Personal Auto  Four-Door Car   
1099                       NaN            NaN            NaN   
2548                       NaN            NaN            NaN   

      Total Claim Amount  
2343                 NaN  
2556                 NaN  
3053                 NaN  
1101                 NaN  
305            76.609295  
1608                 NaN  
2506                 NaN  
263           321.600000  
1099                 NaN  
2548                 NaN

In [12]:
#You should also provide suggestions for fixing any incorrect data types.

#converting to categorcial type
df['ST'] = df['ST'].astype('category')
df['GENDER'] = df['GENDER'].astype('category')
df['Education'] = df['Education'].astype('category')
df['Policy Type'] = df['Policy Type'].astype('category')
df['Vehicle Class'] = df['Vehicle Class'].astype('category')


In [13]:
# converting non categorical values to numeric and removing the '%' for customer lifetime value
df['Customer Lifetime Value'] = pd.to_numeric(df['Customer Lifetime Value'].str.replace('%', ''), errors='coerce')
df['Number of Open Complaints'] = pd.to_numeric(df['Number of Open Complaints'], errors='coerce')


In [14]:
print(df.dtypes)

Customer                       object
ST                           category
GENDER                       category
Education                    category
Customer Lifetime Value       float64
Income                        float64
Monthly Premium Auto          float64
Number of Open Complaints     float64
Policy Type                  category
Vehicle Class                category
Total Claim Amount            float64
dtype: object


In [15]:
#Identify the number of unique values for each column 
df.nunique()

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       0
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

In [16]:
#determine which columns appear to be categorical.
cat = df.select_dtypes(exclude='number')
cat

Customer          ST GENDER             Education     Policy Type  \
0     RB50392  Washington    NaN                Master   Personal Auto   
1     QZ44356     Arizona      F              Bachelor   Personal Auto   
2     AI49188      Nevada      F              Bachelor   Personal Auto   
3     WW63253  California      M              Bachelor  Corporate Auto   
4     GA49547  Washington      M  High School or Below   Personal Auto   
...       ...         ...    ...                   ...             ...   
4003      NaN         NaN    NaN                   NaN             NaN   
4004      NaN         NaN    NaN                   NaN             NaN   
4005      NaN         NaN    NaN                   NaN             NaN   
4006      NaN         NaN    NaN                   NaN             NaN   
4007      NaN         NaN    NaN                   NaN             NaN   

      Vehicle Class  
0     Four-Door Car  
1     Four-Door Car  
2      Two-Door Car  
3               SUV  
4     Four-Door Car  
...             ...  
4003            NaN  
4004            NaN  
4005            NaN  
4006            NaN  
4007            NaN  

[4008 rows x 6 columns]

In [17]:
#You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
df["GENDER"].value_counts()

GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64

Current data has 5 genders as unique values. However as geneder is a binary variable,data requires cleaning

In [19]:
df["ST"].value_counts()

ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

In [20]:
df["Education"].value_counts()

Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64

In [21]:
df["Policy Type"].value_counts()

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

In [22]:
df["Vehicle Class"].value_counts()

Vehicle Class
Four-Door Car    576
Two-Door Car     205
SUV              199
Sports Car        57
Luxury SUV        20
Luxury Car        14
Name: count, dtype: int64

In [23]:
num = df.select_dtypes(include='number')
num

Customer Lifetime Value   Income  Monthly Premium Auto  \
0                         NaN      0.0                1000.0   
1                   697953.59      0.0                  94.0   
2                  1288743.17  48767.0                 108.0   
3                   764586.18      0.0                 106.0   
4                   536307.65  36357.0                  68.0   
...                       ...      ...                   ...   
4003                      NaN      NaN                   NaN   
4004                      NaN      NaN                   NaN   
4005                      NaN      NaN                   NaN   
4006                      NaN      NaN                   NaN   
4007                      NaN      NaN                   NaN   

      Number of Open Complaints  Total Claim Amount  
0                           NaN            2.704934  
1                           NaN         1131.464935  
2                           NaN          566.472247  
3                           NaN          529.881344  
4                           NaN           17.269323  
...                         ...                 ...  
4003                        NaN                 NaN  
4004                        NaN                 NaN  
4005                        NaN                 NaN  
4006                        NaN                 NaN  
4007                        NaN                 NaN  

[4008 rows x 5 columns]

In [24]:
#range of numerical values by retrieving max and min
num.describe()

Customer Lifetime Value        Income  Monthly Premium Auto  \
count             1.068000e+03   1071.000000           1071.000000   
mean              7.936903e+05  39295.701214            193.234360   
std               6.434784e+05  30469.427060           1601.190369   
min               2.004351e+05      0.000000             61.000000   
25%               4.034080e+05  14072.000000             68.000000   
50%               5.881742e+05  36234.000000             83.000000   
75%               8.962872e+05  64631.000000            109.500000   
max               5.816655e+06  99960.000000          35354.000000   

       Number of Open Complaints  Total Claim Amount  
count                        0.0         1071.000000  
mean                         NaN          404.986909  
std                          NaN          293.027260  
min                          NaN            0.382107  
25%                          NaN          202.157702  
50%                          NaN          354.729129  
75%                          NaN          532.800000  
max                          NaN         2893.239678

In [25]:
#Calculating range of three numberical values 
income_r = df['Income'].max() - df['Income'].min()
Monthly_premium_r = df['Monthly Premium Auto'].max() - df['Monthly Premium Auto'].min()
Total_claim_r = df['Total Claim Amount'].max() - df['Total Claim Amount'].min()

print(f"Income range: {income_r}")
print(f"Monthly Premium Auto range: {Monthly_premium_r}")
print(f"Total Claim Amount range: {Total_claim_r}")

Income range: 99960.0
Monthly Premium Auto range: 35293.0
Total Claim Amount range: 2892.857571


Monthly Premium Auto has the highest range among other numerical values.
Monthly Premium Auto has high maximum value and standard deviation. The median premium is 83. Most premiums are between 68 (25th percentile) and 109.5 (75th percentile). 

In [27]:
num = df.select_dtypes('number')
num

Customer Lifetime Value   Income  Monthly Premium Auto  \
0                         NaN      0.0                1000.0   
1                   697953.59      0.0                  94.0   
2                  1288743.17  48767.0                 108.0   
3                   764586.18      0.0                 106.0   
4                   536307.65  36357.0                  68.0   
...                       ...      ...                   ...   
4003                      NaN      NaN                   NaN   
4004                      NaN      NaN                   NaN   
4005                      NaN      NaN                   NaN   
4006                      NaN      NaN                   NaN   
4007                      NaN      NaN                   NaN   

      Number of Open Complaints  Total Claim Amount  
0                           NaN            2.704934  
1                           NaN         1131.464935  
2                           NaN          566.472247  
3                           NaN          529.881344  
4                           NaN           17.269323  
...                         ...                 ...  
4003                        NaN                 NaN  
4004                        NaN                 NaN  
4005                        NaN                 NaN  
4006                        NaN                 NaN  
4007                        NaN                 NaN  

[4008 rows x 5 columns]

In [28]:
#Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns.
#You should also provide your conclusions based on these summary statistics.

df.describe()

Customer Lifetime Value        Income  Monthly Premium Auto  \
count             1.068000e+03   1071.000000           1071.000000   
mean              7.936903e+05  39295.701214            193.234360   
std               6.434784e+05  30469.427060           1601.190369   
min               2.004351e+05      0.000000             61.000000   
25%               4.034080e+05  14072.000000             68.000000   
50%               5.881742e+05  36234.000000             83.000000   
75%               8.962872e+05  64631.000000            109.500000   
max               5.816655e+06  99960.000000          35354.000000   

       Number of Open Complaints  Total Claim Amount  
count                        0.0         1071.000000  
mean                         NaN          404.986909  
std                          NaN          293.027260  
min                          NaN            0.382107  
25%                          NaN          202.157702  
50%                          NaN          354.729129  
75%                          NaN          532.800000  
max                          NaN         2893.239678

In [29]:
#Compute summary statistics for categorical columns and providing your conclusions based on these statistics.
cat.describe()

Customer      ST GENDER Education    Policy Type  Vehicle Class
count      1071    1071    954      1071           1071           1071
unique     1071       8      5         6              3              6
top     RB50392  Oregon      F  Bachelor  Personal Auto  Four-Door Car
freq          1     320    457       324            780            576

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [32]:
customer_location = df['ST'].value_counts()
sorted_locations = customer_location.sort_values(ascending=True)

# top 5 less common locations
top_5_less_common = sorted_locations.head(5)
print(top_5_less_common) 

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64


In [33]:
#another solution using tail
customer_location = df['ST'].value_counts()
sorted_locations = customer_location.sort_values(ascending=False) #states in descending order sorted to retrieve the bottom 5 
top_5_less_common = sorted_locations.tail(5)
print(top_5_less_common)

ST
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64


### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [36]:
Total_policy_sold = df['Policy Type'].value_counts()
print(Total_policy_sold)

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64


In [37]:
max=Total_policy_sold.max()
print(f"Highest number of policies sold is: {max}")

Highest number of policies sold is: 780


In [38]:
#idxmax() method shared during class
# Retrieve the policy type with the highest number of policies sold
most_policy_sold = Total_policy_sold.idxmax()
print(f"The policy type with max sold policies is: {most_policy_sold}")

The policy type with max sold policies is: Personal Auto


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [41]:
personal_auto_policy = df.loc[df['Policy Type'] == 'Personal Auto']
corporate_auto_policy = df.loc[df['Policy Type'] == 'Corporate Auto']

In [42]:
average_income_personal_auto = personal_auto_policy['Income'].mean()
print(f"Personal Auto average: {average_income_personal_auto}")

average_income_corporate_auto = corporate_auto_policy['Income'].mean()
print(f"Corporate Auto average: {average_income_corporate_auto}")

if average_income_corporate_auto > average_income_personal_auto:
    print ("Customers with Personal Auto policy have lower income")
else:
    print("Customers with Corporate Auto policy have lower income")

Personal Auto average: 38180.69871794872
Corporate Auto average: 41390.31196581197
Customers with Personal Auto policy have lower income
